### PR TITLE
Show keyboard on Android entry/editor/searchbar focus

### LINKED ISSUE #5983: Soft Keyboard does not Pop Up when Entry View's Focus is set to True Programmatically
### Description

When Focus is set to an Entry view programmatically (entryTest.Focus()), the cursor blinks in the edit field of the Entry view, but the soft keyboard does not pop up as it should until you physically touch the Entry view.  In comparison to Xamarin Forms where the keyboard pops up when executing entryTest.Focus().

Also, entryTest.Focus() is ignored when placed in the OnAppearing Event.

Video demo of Maui App:

https://user-images.githubusercontent.com/101732760/162827580-c62e0530-9fa1-4f90-b744-90c37caeec57.mp4

Video demo of Xamarin Forms version:

https://user-images.githubusercontent.com/101732760/162827635-9904fcb9-7276-4eba-8d90-ac4c2c7019d9.mp4

XAML:
image

Code Behind:
image

### Steps to Reproduce

1.  Create a new blank Maui app
2.  Add and Entry View and name it entryTest
3.  Add 2 Buttons and set Text to "Set Focus" and "Set Unfocus"
4.  In the Code behind for the 1st button add code entryTest.Focus()
5. In the Code behind for the 2nd button add code enryTest.Unfocus()
6.  Run the App
7.  Click the Set Focus button.
8.  Notice the Entry field cursor blinks
9.  Notice the keyboard does not pop up
10.  Tap the Entry field.
11.  Notice the keyboard pops up.

### Version with bug

Preview 14 (current)

### Last version that worked well

Unknown/Other

### Affected platforms

Android

### Affected platform versions

Android 11

### Did you find any workaround?

Workaround

I added a mapper [here](https://github.com/PureWeen/ShanedlerSamples/blob/main/ShanedlerSamples/Library/Workarounds/EntryWorkarounds.Android.cs) that you can copy to add this behavior for android. 

One thing to note is that if your intention is to open the keyboard you should use
KeyboardManager.ShowKeyboard not `Focus`

Using `Focus` to influence the `Keyboard` has always felt like a hack to me in general. 

### Relevant log output

_No response_

### LINKED ISSUE #13909: Focusing an Entry control fails to open the soft keyboard when targeting net7.0-android
### Description

**Current behavior**
When focusing an Entry control in a Maui app targeting net6.0-android, the text cursor starts blinking in the Entry control AND the soft keyboard is shown.
When doing the same with <TargetFrameworks>net7.0-android</TargetFrameworks>, the soft keyboard is not shown.

**Expected behavior**
The soft keyboard should be shown.

**Environment used for testing**
Visual Studio 2022, v17.4.4
Microsoft Visual Studio Community 2022
Version 17.4.4
VisualStudio.17.Release/17.4.4+33213.308
Microsoft .NET Framework
Version 4.8.04084

Installed Version: Community

Visual C++ 2022   00482-90000-00000-AA347
Microsoft Visual C++ 2022

ASP.NET and Web Tools   17.4.326.54890
ASP.NET and Web Tools

Azure App Service Tools v3.0.0   17.4.326.54890
Azure App Service Tools v3.0.0

Azure Functions and Web Jobs Tools   17.4.326.54890
Azure Functions and Web Jobs Tools

C# Tools   4.4.0-6.22608.27+af1e46ad38d900023f8b1a2839484e471ece1502
C# components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Common Azure Tools   1.10
Provides common services for use by Azure Mobile Services and Microsoft Azure Tools.

Extensibility Message Bus   1.4.1 (main@2ee106a)
Provides common messaging-based MEF services for loosely coupled Visual Studio extension components communication and integration.

Microsoft JVM Debugger   1.0
Provides support for connecting the Visual Studio debugger to JDWP compatible Java Virtual Machines

Mono Debugging for Visual Studio   17.4.19 (8c0a575)
Support for debugging Mono processes with Visual Studio.

NuGet Package Manager   6.4.0
NuGet Package Manager in Visual Studio. For more information about NuGet, visit https://docs.nuget.org/

Razor (ASP.NET Core)   17.0.0.2246202+61cc048d36a3fc9246d2f04625988b19a18ab8f0
Provides languages services for ASP.NET Core Razor.

SQL Server Data Tools   17.0.62207.28050
Microsoft SQL Server Data Tools

TypeScript Tools   17.0.10921.2001
TypeScript Tools for Microsoft Visual Studio

Visual Basic Tools   4.4.0-6.22608.27+af1e46ad38d900023f8b1a2839484e471ece1502
Visual Basic components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Visual F# Tools   17.4.0-beta.22512.4+525d5109e389341bb90b144c24e2ad1ceec91e7b
Microsoft Visual F# Tools

Visual Studio IntelliCode   2.2
AI-assisted development for Visual Studio.

VisualStudio.DeviceLog   1.0
Information about my package

VisualStudio.Mac   1.0
Mac Extension for Visual Studio

VSPackage Extension   1.0
VSPackage Visual Studio Extension Detailed Info

WiX Toolset Visual Studio Extension   1.0.0.18
WiX Toolset Visual Studio Extension version 1.0.0.18
Copyright (c) .NET Foundation and contributors. All rights reserved.

Xamarin   17.4.0.312 (d17-4@be7e8d1)
Visual Studio extension to enable development for Xamarin.iOS and Xamarin.Android.

Xamarin Designer   17.4.0.138 (remotes/origin/d17-4@d36bba3cc9)
Visual Studio extension to enable Xamarin Designer tools in Visual Studio.

Xamarin Templates   17.4.2 (c457c97)
Templates for building iOS, Android, and Windows apps with Xamarin and Xamarin.Forms.

Xamarin.Android SDK   13.1.0.1 (d17-4/13ba222)
Xamarin.Android Reference Assemblies and MSBuild support.
    Mono: a96bde9
    Java.Interop: xamarin/java.interop/d17-4@fcc33ce2
    SQLite: xamarin/sqlite/3.39.3@23e1ae7
    Xamarin.Android Tools: xamarin/xamarin-android-tools/main@0be567a

Xamarin.iOS and Xamarin.Mac SDK   16.1.1.27 (933c6c2c9)
Xamarin.iOS and Xamarin.Mac Reference Assemblies and MSBuild support.

### Steps to Reproduce

1. Use VS to create a new Maui application. In my case it creates a Maui app with this bit in the .csproj file:
     <TargetFrameworks>net6.0-android;net6.0-ios;net6.0-maccatalyst</TargetFrameworks>
I simplified it to 
    <TargetFrameworks>net6.0-android</TargetFrameworks>

2. In MainPage.xml, add a new Entry control:
    <Entry
                    x:Name="NameEntry"
                    BackgroundColor="LightBlue"
                    Loaded="NameEntry_Loaded"
                    Keyboard="Telephone"/>

In MainPage.xaml.cs implement the event handler
    private void NameEntry_Loaded(object sender, EventArgs e)
    {
        NameEntry.Focus();
    }

3. Test.
   - Select Android emulator pixel-5-API-30
   - Observed behavior: the application starts, NameEntry control is focused, soft keyboard shown.

4. Change csproj like this
    <TargetFrameworks>net7.0-android</TargetFrameworks>

5. Test. Observed behavior in ERROR: the application starts, NameEntry control is focused, soft keyboard NOT SHOWN.

### Link to public reproduction project repository

N/A

### Version with bug

6.0.312

### Last version that worked well

6.0.312

### Affected platforms

Android

### Affected platform versions

Android 11

### Did you find any workaround?

No
Tried using different event handlers, tried instantiating Android InputMethodManager and opening the keyboard directly, tried setting IsEnabled to false then back to true, tried Unfocus() then Focus(). The only change that had any effect was setting Target to net6.0.
See the steps and code examples in this StackOverflow question: https://stackoverflow.com/questions/75502572/how-to-use-getsystemservice-in-maui

### Relevant log output

```shell
N/A
```

### PR DESCRIPTION
**Pull request by PureWeen**
### Description of Change

Show keyboard on Android entry/editor/searchbar focus

Re-do of https://github.com/dotnet/maui/pull/12890 with tweaks to ensure passing tests on CI

### Issues Fixed

Fixes #5983
Fixes #13909

### CODE REVIEW

File: src/Core/src/Platform/Android/KeyboardManager.cs [hunk 1]
```diff
@@ -107,12 +92,22 @@ void ShowKeyboard()
...
 				return false;
 
 			var result = insets.IsVisible(WindowInsetsCompat.Type.Ime());
-			return result;
+			var size = insets.GetInsets(WindowInsetsCompat.Type.Ime());
+
+			if (!result && size.Top == 0 && size.Bottom == 0)
+				return false;
+
+			return result && (size.Top > 0 || size.Bottom > 0);
```
Comment (mandel-macaque) [inline src/Core/src/Platform/Android/KeyboardManager.cs:110]: Can this two be combined in a single result?

File: src/Core/src/Handlers/Editor/EditorHandler.Android.cs [hunk 2]
```diff
@@ -108,6 +108,12 @@ public static void MapPlaceholderColor(IEditorHandler handler, IEditor editor)
 		public static void MapSelectionLength(IEditorHandler handler, ITextInput editor) =>
 			handler.PlatformView?.UpdateSelectionLength(editor);
 
+		static void MapFocus(IEditorHandler handler, IEditor editor, object? args)
+		{
+			if (args is FocusRequest request)
+				handler.PlatformView?.Focus(request);
```
Comment (mandel-macaque) [inline src/Core/src/Handlers/Editor/EditorHandler.Android.cs:114]: SI the focus method the one that was added in the extension? That one already checks for null right? Could we pass it as 
```suggestion
			handler.PlatformView?.Focus(args is FocusRequest);
```

File: src/Core/src/Handlers/Entry/EntryHandler.Android.cs [hunk 3]
```diff
@@ -128,6 +128,12 @@ public static void MapClearButtonVisibility(IEntryHandler handler, IEntry entry)
 				handler.PlatformView?.UpdateClearButtonVisibility(entry, platformHandler.GetClearButtonDrawable);
 		}
 
+		static void MapFocus(IEntryHandler handler, IEntry entry, object? args)
+		{
+			if (args is FocusRequest request)
+				handler.PlatformView?.Focus(request);
```
Comment (mandel-macaque) [inline src/Core/src/Handlers/Entry/EntryHandler.Android.cs:134]: Same question

**Review by mandel-macaque — COMMENTED** [review]
Some questions

File: src/Core/src/Platform/Android/KeyboardManager.cs [hunk 4]
```diff
@@ -101,13 +101,11 @@ public static bool IsSoftKeyboardVisible(this AView view)
...
-			var result = insets.IsVisible(WindowInsetsCompat.Type.Ime());
 			var size = insets.GetInsets(WindowInsetsCompat.Type.Ime());
+			var isVisible = insets.IsVisible(WindowInsetsCompat.Type.Ime());
+			var hasSize = size.Top > 0 || size.Bottom > 0;
 
-			if (!result && size.Top == 0 && size.Bottom == 0)
-				return false;
-
-			return result && (size.Top > 0 || size.Bottom > 0);
+			return hasSize || isVisible;
```
Comment (mandel-macaque) [inline src/Core/src/Platform/Android/KeyboardManager.cs:108]: <3

**Review by mandel-macaque — APPROVED** [review]

**Review by rachelkang — APPROVED** [review]

**Comment by rachelkang** [comment]
backport was approved via #12890

**Pull request merged**